### PR TITLE
WIP: low-level envelopes code

### PR DESCRIPTION
**Pull request by dengemann**
This is the first step in a series of planned commits and contributions with the following schedule.

1. Numpy API for computing power envelopes in a given epoch (this PR)
2. Add resting state data to mne.datasets to be able to make an example
3. Add high-level label API based on the MNE-C `mne_compute_raw_inverse` to allow using Raw and Epochs for representing source time courses.
4. Make actual example
5. Add more options for orthogonalization (via @dokato)

@agramfort @larsoner would you have suggestion how to organize this schedule in terms of separate PRs. In principle we could do everything here. It will just be quite a bit of content. Perhaps 2, 3 and 5 could be separate PRs. Then the dependence of this PR to be mergeable would be having 2 in.

cc @dokato @agramfort @SherazKhan

### CODE REVIEW

**Comment by dengemann** [comment]
As a result we should then be able to have this in MNE:

screenshot 2018-08-02 18 29 31

File: mne/connectivity/spectral/envelope.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    """Correlate 2 matrices along last axis.
+
+    Parameters
+    ----------
+    x : np.ndarray of shape(n_time_series, n_times)
+        The first set of vectors.
+    y : np.ndarray of shape(n_time_series, n_times)
+        The second set of vectors.
+
+    Retrurns
```
Comment (jasmainak) [inline mne/connectivity/spectral/envelope.py]: dengmannisms :-)

**Comment by dengemann** [comment]
Just my way of saying « hi I’m back » :)

**Comment by jona-sassenhagen** [comment]
>Just my way of saying « hi I’m back » :)

I'll only believe it when I see it :)

File: mne/connectivity/spectral/envelope.py (hunk 1 above)
Comment (jona-sassenhagen) [reply mne/connectivity/spectral/envelope.py]: Shouldn't this one be more generally available? Under mne.utils maybe, right by mne.utils.compute_corr?

Comment (dengemann) [reply mne/connectivity/spectral/envelope.py]: Ja we can deal with it later. Now we have other obstacles in moving this forward.

Comment (jona-sassenhagen) [reply mne/connectivity/spectral/envelope.py]: I.e. the retrurns stay :D

**Comment by larsoner** [comment]
@dengemann this is the code I would expect to run something like:
```
raw.load_data().filter(8, 12).apply_hilbert()
events = mne.make_fixed_length_events(raw, duration=1.)
epochs = mne.Epochs(raw, events, tmin=0, tmax=1. - 1. / raw.info['sfreq'],
                    decim=10, baseline=None, preload=False)
stcs = mne.minimum_norm.apply_inverse_epochs(epochs, inv)
conn = mne.connectivity.envelope_correlation(stcs)
```
I think this should be doable from a coding perspective, and shouldn't take much memory since you can compute the correlation iteratively.

Once #5514 is in I'm happy to help with this since I know some folks who would be interested in doing these computations.

**Comment by dengemann** [comment]
@larsoner 

```
# assume we have defined labels already
raw.load_data().filter(8, 12).apply_hilbert()
events = mne.make_fixed_length_events(raw, duration=1.)
epochs = mne.Epochs(raw, events, tmin=0, tmax=1. - 1. / raw.info['sfreq'],
                    decim=10, baseline=None, preload=False)
stcs = mne.minimum_norm.apply_inverse_epochs(epochs, inv)
conn = mne.connectivity.envelope_correlation(stcs, labels=labels)
assert len(conn) == len(labels)
```

I'd also make a flexible API that accepts STCs or numpy arrays as inputs.

**Comment by larsoner** [comment]
I *think* that this should be doable within the existing API using generators. Maybe not with the orthogonalization step, though, depending on where/how that is implemented. Will have to think about it / try it out.

If you have working code for this that you want me to try to look and and adapt / translate at some point let me know.

**Comment by dengemann** [comment]
> I think that this should be doable within the existing API using generators. Maybe not with the orthogonalization step, though, depending on where/how that is implemented. Will have to think about it / try it out.

The orthogonalization is working well (inside this function) if you have a few labels only. We'll have to see what is most convenient.

> If you have working code for this that you want me to try to look and and adapt / translate at some point let me know.

I am happy to do it / make the start once we have satisfying resting state examples. Let's tackle #5692 and see where it gets us. Then we make a separate power envelope example where all comes together.

**Comment by mmagnuski** [comment]
@dengemann Just reminding that #5514 #5692 are merged :)

**Pull request closed without merging**